File: Hackerrank/minimum-swaps-2.py
# Minimum Swaps 2: https://www.hackerrank.com/challenges/minimum-swaps-2/problem

#!/bin/python3

import math
import os
import random
import re
import sys

# Complete the minimumSwaps function below.

def swap(a, b):
    tmp = a
    a = b
    b = tmp

def minimumSwaps(arr):
    largest = len(arr)
    swaps = 0
    i = 0
    while i < largest:
        if arr[i] == i + 1:
            i += 1
            continue
        loc = arr[i] - 1
        arr[i], arr[loc] = arr[loc], arr[i]
        swaps += 1
    return swaps

if __name__ == '__main__':
    fptr = open(os.environ['OUTPUT_PATH'], 'w')

    n = int(input())

    arr = list(map(int, input().rstrip().split()))

    res = minimumSwaps(arr)

    fptr.write(str(res) + '\n')

    fptr.close()

File: Hackerrank/MinimumSwaps2.java
import java.io.*;
import java.math.*;
import java.security.*;
import java.text.*;
import java.util.*;
import java.util.concurrent.*;
import java.util.regex.*;

public class MinimumSwaps2 {

    // Complete the minimumSwaps function below.
    static int minimumSwaps(int[] arr) {
        int counter = 0;
        for (int i = 0; i < arr.length; i++) {
            int expectedPosition = arr[i] - 1;
            int currentPosition = i;
            if (expectedPosition - currentPosition < 0) {
                counter++;
                int temp = arr[currentPosition];
                arr[currentPosition] = arr[expectedPosition];
                arr[expectedPosition] = temp;
                i--;
            }
        }
        System.out.println(counter);
        return counter;
    }

    private static final Scanner scanner = new Scanner(System.in);

    public static void main(String[] args) throws IOException {
        BufferedWriter bufferedWriter = new BufferedWriter(new FileWriter(System.getenv("OUTPUT_PATH")));

        int n = scanner.nextInt();
        scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");

        int[] arr = new int[n];

        String[] arrItems = scanner.nextLine().split(" ");
        scanner.skip("(\r\n|[\n\r\u2028\u2029\u0085])?");

        for (int i = 0; i < n; i++) {
            int arrItem = Integer.parseInt(arrItems[i]);
            arr[i] = arrItem;
        }

        int res = minimumSwaps(arr);

        bufferedWriter.write(String.valueOf(res));
        bufferedWriter.newLine();

        bufferedWriter.close();

        scanner.close();
    }
}
